File: artifacts/api-server/src/routes/learn.ts
import { Router, type IRouter } from "express";
import { sql } from "drizzle-orm";
import { db, documents, documentChunks, learnedAnswers } from "../lib/db";
import { requireJwt } from "../middlewares/auth";
import { embedOne, toVectorLiteral } from "../lib/embeddings";

const router: IRouter = Router();

// Persist a Q&A as a single chunk so RAG retrieval can surface it later.
router.post("/learn", requireJwt, async (req, res) => {
  const tenantId = req.auth!.tid;
  const question = (req.body?.question || "").toString().trim();
  const answer = (req.body?.answer || "").toString().trim();
  const sourceHandoffId = req.body?.source_handoff_id
    ? Number(req.body.source_handoff_id)
    : null;
  if (!question || !answer) return res.status(400).json({ detail: "question and answer required" });

  const [learned] = await db
    .insert(learnedAnswers)
    .values({ tenantId, question, answer, sourceHandoffId })
    .returning();

  const [doc] = await db
    .insert(documents)
    .values({
      tenantId,
      name: `Learned: ${question.slice(0, 60)}`,
      type: "learned",
      sizeBytes: Buffer.byteLength(question + answer),
      status: "Indexed",
      source: `learned-${learned.id}`,
    })
    .returning();

  const text = `Q: ${question}\nA: ${answer}`;
  const emb = await embedOne(text);
  if (emb) {
    await db.execute(sql`
      INSERT INTO document_chunks (tenant_id, document_id, chunk_index, content, embedding)
      VALUES (${tenantId}, ${doc.id}, 0, ${text}, ${toVectorLiteral(emb)}::vector)
    `);
  } else {
    await db.insert(documentChunks).values({
      tenantId,
      documentId: doc.id,
      chunkIndex: 0,
      content: text,
    });
  }
  return res.json({ ok: true, id: learned.id, documentId: doc.id });
});

export default router;

File: artifacts/api-server/app/routes/learn.py
from typing import Annotated
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth import get_current_user
from app.database import get_db
from app.models import Document, DocumentChunk, LearnedAnswer
from app.schemas import LearnRequest, LearnResponse
from app.services.embeddings import embed_one

router = APIRouter()


@router.post("/learn", response_model=LearnResponse)
async def learn(
    body: LearnRequest,
    auth: Annotated[dict, Depends(get_current_user)],
    db: AsyncSession = Depends(get_db),
):
    learned = LearnedAnswer(
        tenant_id=auth["tid"],
        question=body.question,
        answer=body.answer,
        source_handoff_id=body.source_handoff_id,
    )
    db.add(learned)
    await db.flush()

    text = f"Q: {body.question}\nA: {body.answer}"
    doc = Document(
        tenant_id=auth["tid"],
        name=f"learned-{learned.id}",
        type="learned",
        size_bytes=len(text.encode()),
        status="indexed",
        source=f"learned-{learned.id}",
    )
    db.add(doc)
    await db.flush()

    embedding = await embed_one(text)
    db.add(DocumentChunk(
        tenant_id=auth["tid"],
        document_id=doc.id,
        chunk_index=0,
        content=text,
        embedding=embedding,
    ))

    await db.commit()
    return LearnResponse(ok=True, id=learned.id, document_id=doc.id)
